packages/ai-bot/helpers.ts

```diff
@@ -137,6 +137,61 @@ export async function getPromptParts(
   };
 }
 
+function getAggregatedReplacement(event: IRoomEvent) {
+  /**
+   * When replacements are applied, the server aggregates
+   * them into a single event.
+   *
+   * The latest version is placed within the unsigned
+   * section of the event.
+   *
+   * Here we extract any replacement and return the
+   * latest version, but with *the original id*
+   */
+  let finalRawEvent: IRoomEvent;
+  const originalEventId = event.event_id;
+  let replacedRawEvent: IRoomEvent =
+    event.unsigned?.['m.relations']?.['m.replace'];
+  if (replacedRawEvent) {
+    finalRawEvent = replacedRawEvent;
+    finalRawEvent.event_id = originalEventId;
+  } else {
+    finalRawEvent = event;
+  }
+  return finalRawEvent;
+}
+
+function applyAllReplacements(eventlist: IRoomEvent[]): IRoomEvent[] {
+  // First apply any server-side aggregations
+  let eventsWithAggregatedReplacements = eventlist.map(
+    getAggregatedReplacement,
+  );
+  // Now if the event list we have doesn't have aggregations but still
+  // has replacements, we need to apply them manually
+  let eventsMap = new Map<string, IRoomEvent>();
+  for (let event of eventsWithAggregatedReplacements) {
+    let canonicalEventId;
+    if (event.content['m.relates_to']?.rel_type === 'm.replace') {
```

Comment:
Will this ever happen once we are filtering m.replace events out?

Reply:
No, it shouldn't. 

We could enforce timing and do the infra piece first, and go through updating the tests to record new versions with the right structure, then we could leave this piece out.

Reply:
Let's just add a comment that this is temporary, for backward compat while we update infra and tests